File: transparent/src/main/java/zhan/transparent/TransparentDelegate.java
package zhan.transparent;

import android.graphics.Color;
import android.view.View;
import android.view.ViewTreeObserver;
import java.util.ArrayList;
import java.util.List;


public class TransparentDelegate
    implements ViewTreeObserver.OnGlobalLayoutListener, ITransparentDelegate {

  private static final int MIN_FRACTION = 0;
  private static final int MAX_FRACTION = 1;

  private View mRootView;
  private List<OnTransparentListener> mOnTransparentListeners = new ArrayList<>();

  private float mMaxOffset;
  private int mNormalGroundColor;
  private int mBackGroundColor;

  public TransparentDelegate(View rootView) {
    if (rootView == null) {
      throw new RuntimeException("TransparentDelegate root view is null");
    }
    mRootView = rootView;
    mNormalGroundColor = mRootView.getResources().getColor(android.R.color.transparent);
    mBackGroundColor = mNormalGroundColor;
    mRootView.getViewTreeObserver().removeGlobalOnLayoutListener(this);
  }

  @Override public void onGlobalLayout() {
    mRootView.setBackgroundColor(mBackGroundColor);
    mRootView.getViewTreeObserver().removeGlobalOnLayoutListener(this);
  }

  public void addOnScrollStateListener(OnTransparentListener listener) {
    if (listener != null) {
      if (!mOnTransparentListeners.contains(listener)) {
        mOnTransparentListeners.add(listener);
      }
    }
  }

  public void removeOnScrollStateListener(OnTransparentListener listener) {
    if (listener != null) {
      if (mOnTransparentListeners.contains(listener)) {
        mOnTransparentListeners.remove(listener);
      }
    }
  }

  public void removeOnScrollStateListenerAll() {
    mOnTransparentListeners.clear();
  }

  public int changeAlpha(int color, float fraction) {
    int red = Color.red(color);
    int green = Color.green(color);
    int blue = Color.blue(color);
    int alpha = (int) (Color.alpha(color) * fraction);
    return Color.argb(alpha, red, green, blue);
  }

  @Override public void updateTop(float top) {
    if (mMaxOffset <= 0 && mBackGroundColor == mNormalGroundColor) {
      return;
    }
    float fraction = top / mMaxOffset;

    if (fraction <= MIN_FRACTION) {
      fraction = MIN_FRACTION;
      for (OnTransparentListener listener : mOnTransparentListeners) {
        listener.onTransparentStart(fraction);
      }
    }
    if (fraction >= MAX_FRACTION) {
      fraction = MAX_FRACTION;
      for (OnTransparentListener listener : mOnTransparentListeners) {
        listener.onTransparentEnd(fraction);
      }
    }
    int newColor = changeAlpha(mBackGroundColor, fraction);
    mRootView.setBackgroundColor(newColor);

    for (OnTransparentListener listener : mOnTransparentListeners) {
      listener.onTransparentUpdateFraction(fraction);
    }
  }

  @Override public void setMaxOffset(float offset) {
    this.mMaxOffset = offset;
  }

  @Override public void setColorToBackGround(int backGroundColor) {
    this.mBackGroundColor = backGroundColor;
  }

  public float getOffset() {
    return mMaxOffset;
  }

  public int getColorToBackGround() {
    return mBackGroundColor;
  }
}

File: transparent/src/main/java/zhan/transparent/TransparentDelegate.kt
package zhan.transparent

import android.graphics.Color
import android.view.View
import android.view.ViewTreeObserver
import androidx.core.content.ContextCompat
import java.util.*


class TransparentDelegate(private val mRootView: View) : ViewTreeObserver.OnGlobalLayoutListener,
        ITransparentDelegate {
    private val mOnTransparentListeners = ArrayList<OnTransparentListener>()

    private var offset: Float = 0.toFloat()
    private val mNormalGroundColor: Int = ContextCompat.getColor(mRootView.context,
            android.R.color.transparent)
    private var mBackGroundColor: Int = 0

    init {
        mBackGroundColor = mNormalGroundColor
        mRootView.viewTreeObserver.removeGlobalOnLayoutListener(this)
    }

    override fun onGlobalLayout() {
        mRootView.setBackgroundColor(mBackGroundColor)
        mRootView.viewTreeObserver.removeGlobalOnLayoutListener(this)
    }

    fun addOnScrollStateListener(listener: OnTransparentListener?) {
        if (listener != null) {
            if (!mOnTransparentListeners.contains(listener)) {
                mOnTransparentListeners.add(listener)
            }
        }
    }

    fun removeOnScrollStateListener(listener: OnTransparentListener?) {
        if (listener != null) {
            if (mOnTransparentListeners.contains(listener)) {
                mOnTransparentListeners.remove(listener)
            }
        }
    }

    fun removeOnScrollStateListenerAll() {
        mOnTransparentListeners.clear()
    }

    private fun changeAlpha(color: Int, fraction: Float): Int {
        val red = Color.red(color)
        val green = Color.green(color)
        val blue = Color.blue(color)
        val alpha = (Color.alpha(color) * fraction).toInt()
        return Color.argb(alpha, red, green, blue)
    }

    override fun updateTop(top: Float) {
        if (offset <= 0 && mBackGroundColor == mNormalGroundColor) {
            return
        }
        var fraction = top / offset

        if (fraction <= MIN_FRACTION) {
            fraction = MIN_FRACTION.toFloat()
            for (listener in mOnTransparentListeners) {
                listener.onTransparentStart(fraction)
            }
        }
        if (fraction >= MAX_FRACTION) {
            fraction = MAX_FRACTION.toFloat()
            for (listener in mOnTransparentListeners) {
                listener.onTransparentEnd(fraction)
            }
        }
        val newColor = changeAlpha(mBackGroundColor, fraction)
        mRootView.setBackgroundColor(newColor)

        for (listener in mOnTransparentListeners) {
            listener.onTransparentUpdateFraction(fraction)
        }
    }

    override fun setMaxOffset(offset: Float) {
        this.offset = offset
    }

    override fun setColorToBackGround(backGroundColor: Int) {
        this.mBackGroundColor = backGroundColor
    }

    fun getColorToBackGround(): Int {
        return mBackGroundColor
    }

    companion object {

        private const val MIN_FRACTION = 0
        private const val MAX_FRACTION = 1
    }
}
